deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/layers/convolution/Deconvolution2DLayer.java

```diff
@@ -0,0 +1,288 @@
+
+        if (convolutionMode == ConvolutionMode.Same) {
+            throw new IllegalArgumentException("Border mode Same currently not supported.");
+        }
+
+        INDArray bias = getParamWithNoise(DeconvolutionParamInitializer.BIAS_KEY, training);
+        INDArray weights = getParamWithNoise(DeconvolutionParamInitializer.WEIGHT_KEY, training);
+
+        //Input validation: expect rank 4 matrix
+        if (input.rank() != 4) {
+            String layerName = conf.getLayer().getLayerName();
+            if (layerName == null)
+                layerName = "(not named)";
+            throw new DL4JInvalidInputException("Got rank " + input.rank()
+                    + " array as input to SeparableConvolution2D (layer name = " + layerName + ", layer index = "
+                    + index + ") with shape " + Arrays.toString(input.shape()) + ". "
+                    + "Expected rank 4 array with shape [minibatchSize, layerInputDepth, inputHeight, inputWidth]."
+                    + (input.rank() == 2
+                    ? " (Wrong input type (see InputType.convolutionalFlat()) or wrong data type?)"
+                    : "")
+                    + " " + layerId());
+        }
+
+        int inDepth = weights.size(0);
+        int outDepth = weights.size(1);
+
+        if (input.size(1) != inDepth) {
+            String layerName = conf.getLayer().getLayerName();
+            if (layerName == null)
+                layerName = "(not named)";
+            throw new DL4JInvalidInputException("Cannot do forward pass in Deconvolution2D layer (layer name = " + layerName
+                    + ", layer index = " + index + "): input array depth does not match CNN layer configuration"
+                    + " (data input depth = " + input.size(1) + ", [minibatch,inputDepth,height,width]="
+                    + Arrays.toString(input.shape()) + "; expected" + " input depth = " + inDepth + ") "
+                    + layerId());
+        }
+        int kH = weights.size(2);
+        int kW = weights.size(3);
+
+        int[] dilation = layerConf().getDilation();
+        int[] kernel = layerConf().getKernelSize();
+        int[] strides = layerConf().getStride();
+
+        int[] pad;
+        int[] outSize;
+        if (convolutionMode == ConvolutionMode.Same) {
+            outSize = ConvolutionUtils.getDeconvolutionOutputSize(input, kernel, strides, null, convolutionMode, dilation); //Also performs validation
+            pad = ConvolutionUtils.getSameModeTopLeftPadding(outSize, new int[] {input.size(2), input.size(3)}, kernel,
+                    strides, dilation );
+        } else {
+            pad = layerConf().getPadding();
+            outSize = ConvolutionUtils.getDeconvolutionOutputSize(input, kernel, strides, pad, convolutionMode, dilation); //Also performs validation
+        }
+
+        int outH = outSize[0];
+        int outW = outSize[1];
+
+
+        if (helper != null) {
```

Comment:
Does CuDNN support deconv? If so, we need separate tests for it.
If not, we should probably override initializeHelper() and make it a no-op.